### PR TITLE
[openweathermap] Add support for One Call API #7308

### PR DESCRIPTION
**Pull request by Wolfgang1966**
Signed-off-by: wolfii <wolfgang.klimt@consol.de>

This is the OH3 version of PR #8395

It adds support for the Openweathermap One Call API to the openweathermap binding of @cweitkamp  as requested in #7308 . The binding is compatible to the existing version but adds two new things using the free version of the One Call API. For forecasts it should be simple to adapt the configuration from the existing thing to the new thing as the binding uses the same channel names where possible. Additionally a new thing allows to retrieve historical data for the last (up to 5) days.

Nearly all review comments that I got for the 2.5 version of the binding are resolved also in this version. For compatibility I will add the temperature channel as suggested by @boehan . Review comments are welcome. Please be patient, it may take several days until I find the time to respond.

Find a snapshot build and example configs for testing here:

https://home.klimt.de/owncloud/index.php/s/Y7P4mWfXJs7iWNW

### CODE REVIEW

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/OpenWeatherMapBindingConstants.java [hunk 1]
```diff
@@ -49,38 +53,66 @@
...
     public static final String CHANNEL_SNOW = "snow";
     public static final String CHANNEL_VISIBILITY = "visibility";
     public static final String CHANNEL_UVINDEX = "uvindex";
+    public static final String CHANNEL_PRECIPITATION = "precipitation";
 
     // List of all configuration
     public static final String CONFIG_FORECAST_DAYS = "forecastDays";
+
+    // relative number of days in history
+    public static final String CONFIG_HISTORY_DAYS = "historyDays";
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/OpenWeatherMapBindingConstants.java:117]: Dow we need this constant?

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/connection/OpenWeatherMapConnection.java [hunk 2]
```diff
@@ -38,8 +39,10 @@
 import org.openhab.binding.openweathermap.internal.dto.OpenWeatherMapJsonHourlyForecastData;
 import org.openhab.binding.openweathermap.internal.dto.OpenWeatherMapJsonUVIndexData;
 import org.openhab.binding.openweathermap.internal.dto.OpenWeatherMapJsonWeatherData;
+import org.openhab.binding.openweathermap.internal.dto.onecall.OpenWeatherMapOneCallAPIData;
+import org.openhab.binding.openweathermap.internal.dto.onecallhist.OpenWeatherMapOneCallHistAPIData;
 import org.openhab.binding.openweathermap.internal.handler.OpenWeatherMapAPIHandler;
-import org.openhab.core.cache.ByteArrayFileCache;
+import org.openhab.binding.openweathermap.internal.utils.ByteArrayFileCache;
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/connection/OpenWeatherMapConnection.java:45]: The `ByteArrayFileCache``is now available in OHC. Please revert this change.

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/utils/ByteArrayFileCache.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.openweathermap.internal.utils;
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/utils/ByteArrayFileCache.java:13]: Please remove this class. It is not needed anymore because it is available in OHC.

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Weather_.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.openweathermap.internal.dto.onecall;
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Weather_.java:13]: What is the difference between this class and the other `Weather` class? (see https://github.com/openhab/openhab-addons/pull/9062/files#diff-8d98c5ea97753b200daab2546b605d6713a25af0a2c16ef830cc678bc2eb35b2)

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Weather__.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.openweathermap.internal.dto.onecall;
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Weather__.java:13]: What is the difference between this class and the other `Weather` class? (see https://github.com/openhab/openhab-addons/pull/9062/files#diff-8d98c5ea97753b200daab2546b605d6713a25af0a2c16ef830cc678bc2eb35b2)

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecallhist/Weather_.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.openweathermap.internal.dto.onecallhist;
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecallhist/Weather_.java:13]: What is the difference between this class and the other `Weather` class? (see https://github.com/openhab/openhab-addons/pull/9062/files#diff-9a8c2612d60ade9bb74e96978e5aba5d07bd90f08579e2001039c9c212e79cd7)

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap_de.properties [hunk 7]
```diff
@@ -4,11 +4,11 @@ binding.openweathermap.description = OpenWeatherMap - Aktuelles Wetter und Progn
 
 # bridge types
 thing-type.openweathermap.weather-api.label = OpenWeatherMap Konto
-thing-type.openweathermap.weather-api.description = Erm�glicht den Zugriff auf die OpenWeatherMap API.
+thing-type.openweathermap.weather-api.description = Erm�glicht den Zugriff auf die OpenWeatherMap API.
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap_de.properties:7]: Can you check the file encoding? It needs to be ISO-8859.

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap_fr.properties [hunk 8]
```diff
@@ -1,257 +1,261 @@
 # binding
 binding.openweathermap.name = Extension OpenWeatherMap
-binding.openweathermap.description = L'extension OpenWeatherMap fournit la m�t�o actuelle et les pr�visions dans votre ville.
+binding.openweathermap.description = L'extension OpenWeatherMap fournit la m�t�o actuelle et les pr�visions dans votre ville.
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap_fr.properties:3]: see above (file encoding).

**Review by cweitkamp — CHANGES_REQUESTED** [review]
Thanks for porting this to openHAB 3. I left some final comments.

**Comment by rgrollfitz** [comment]
Hi @Wolfgang1966 

thanks for your efforts on the development of this binding - works great so far in OH3. Is it possible to also include 'conditionID' and the 'icon' (the ID not the translated svg image) as a channel of the onecall part of this binding?

The 'conditionID' is very handy when it comes to reacting on different wether conditions e.g. in a rule, due to ist shortness.

The 'icon' currently got tranlated internally to an icon data-attribute as far as I understand.
To create custom widgets it would be helpfull if there might be a way, accessing the original icon-ID from the API answer. The benefit of the icon name over the conditionID is, that there is a differentiation between day and night at the last character of the name, which would be helpfull for some cases:

```
            "weather": [
                   {
                   "id": 801,
                   "main": "Clouds",
                   "description": "few clouds",
                   "icon": "02n"
                   }
            ],
```
Thanks again for your work!

**Comment by Wolfgang1966** [comment]
> Hi @Wolfgang1966
> 
> thanks for your efforts on the development of this binding - works great so far in OH3. Is it possible to also include 'conditionID' and the 'icon' (the ID not the translated svg image) as a channel of the onecall part of this binding?
> 
> The 'conditionID' is very handy when it comes to reacting on different wether conditions e.g. in a rule, due to ist shortness.
> 
> The 'icon' currently got tranlated internally to an icon data-attribute as far as I understand.
> To create custom widgets it would be helpfull if there might be a way, accessing the original icon-ID from the API answer. The benefit of the icon name over the conditionID is, that there is a differentiation between day and night at the last character of the name, which would be helpfull for some cases:
> 
> ```
>             "weather": [
>                    {
>                    "id": 801,
>                    "main": "Clouds",
>                    "description": "few clouds",
>                    "icon": "02n"
>                    }
>             ],
> ```
> 
> Thanks again for your work!

Dear @rgrollfitz , condition-id is already there. Check for channel condition-id. it is just marked as advanced, so it is not probably not shown in the UI. The decision to mark it as advanced seems to have been made by @cweitkamp who developed the original binding. If he agrees, I can change that, but for sure there is a reason behind. Could be that the binding already provides such a lot of different things that there was the need for some simplification.

**Comment by rgrollfitz** [comment]
> Dear @rgrollfitz , condition-id and icon_id are already there, they are just marked as advanced, so probably not shown in the UI. The decision to mark them as advanced seems to have been made by @cweitkamp who developed the original binding. If he agrees, I can change that, but for sure there is a reason behind. Could be that the binding already provides such a lot of different things that there was the need for some simplification.

Awesome, it works flawlessly. Even the iconID is already there. TY!

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/OpenWeatherMapBindingConstants.java (hunk 1 above)
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/OpenWeatherMapBindingConstants.java]: Nope :-) 
Removed

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/connection/OpenWeatherMapConnection.java (hunk 2 above)
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/connection/OpenWeatherMapConnection.java]: Done. I got ClassNotFoundExceptions for my last update, this is why I added it originally. Looks like my 3.0 snapshot was too old then.

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/utils/ByteArrayFileCache.java (hunk 3 above)
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/utils/ByteArrayFileCache.java]: Done

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Weather_.java (hunk 4 above)
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Weather_.java]: This was what came out of the generated json mapper, didn't look into it in detail. However, works also when using the Weather class  in all places. Removed it

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Weather__.java (hunk 5 above)
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecall/Weather__.java]: Also removed

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecallhist/Weather_.java (hunk 6 above)
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/dto/onecallhist/Weather_.java]: Also removed

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap_de.properties (hunk 7 above)
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap_de.properties]: corrected encoding (hopefully :-) )

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap_fr.properties (hunk 8 above)
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/i18n/openweathermap_fr.properties]: see above :-)

File: bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/binding/binding.xml [hunk 9]
```diff
@@ -5,6 +5,6 @@
 
 	<name>OpenWeatherMap Binding</name>
 	<description>OpenWeatherMap - Current weather and forecasts in your city.</description>
-	<author>Christoph Weitkamp</author>
+	<author>Christoph Weitkamp, Wolfgang Klimt</author>
```
Comment (cweitkamp) [inline bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/binding/binding.xml:8]: One final comment from my side. The `author`-tag now is deprecated. Please remove it. Thanks for your efforts.

Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/resources/OH-INF/binding/binding.xml]: Done.

**Review by cweitkamp — APPROVED** [review]
LGTM.Thanks.

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/OpenWeatherMapOneCallHandler.java [hunk 10]
```diff
@@ -0,0 +1,636 @@
...
+                } else {
+                    if (forecastDays == 0 && newForecastDays > 0) {
+                        toBeAddedChannels.addAll(createChannelsForGroup(CHANNEL_GROUP_FORECAST_TODAY,
+                                CHANNEL_GROUP_TYPE_ONECALL_DAILY_FORECAST));
+                    }
+                    if (forecastDays <= 1 && newForecastDays > 1) {
+                        toBeAddedChannels.addAll(createChannelsForGroup(CHANNEL_GROUP_FORECAST_TOMORROW,
+                                CHANNEL_GROUP_TYPE_ONECALL_DAILY_FORECAST));
+                    }
+                    for (int i = (forecastDays < 2) ? 2 : forecastDays; i < newForecastDays; ++i) {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/OpenWeatherMapOneCallHandler.java:164]:
```suggestion
                    for (int i = Math.max(2, forecastDays); i < newForecastDays; i++) {
```

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/AbstractOpenWeatherMapHandler.java [hunk 11]
```diff
@@ -205,8 +206,10 @@ protected State getQuantityTypeState(@Nullable Number value, Unit<?> unit) {
     }
 
     protected List<Channel> createChannelsForGroup(String channelGroupId, ChannelGroupTypeUID channelGroupTypeUID) {
-        logger.debug("Building channel group '{}' for thing '{}'.", channelGroupId, getThing().getUID());
+        logger.debug("Building channel group '{}' for thing '{}' and GroupType '{}'.", channelGroupId,
+                getThing().getUID(), channelGroupTypeUID);
         List<Channel> channels = new ArrayList<>();
+        @Nullable
```
[the remark this answers is not in the document]
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/AbstractOpenWeatherMapHandler.java]: Removed

File: bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/OpenWeatherMapOneCallHandler.java (hunk 10 above)
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.openweathermap/src/main/java/org/openhab/binding/openweathermap/internal/handler/OpenWeatherMapOneCallHandler.java]: Changed.

**Review by cpmeister — APPROVED** [review]
